capa/rules/__init__.py

```diff
@@ -1401,154 +1423,320 @@ def __getitem__(self, rulename):
     def __contains__(self, rulename):
         return rulename in self.rules
 
+    # unstable
     @staticmethod
-    def _index_rules_by_feature(rules) -> Tuple[Dict[Feature, Set[str]], List[str]]:
+    def _score_feature(scores_by_rule: Dict[str, int], node: capa.features.common.Feature) -> int:
+        """
+        Score the given feature by how "uncommon" it is, where a higher score is more uncommon.
+        Features that are not good for indexing will have a low score, or 0.
         """
-        split the given rules into two structures:
-          - "easy rules" are indexed by feature,
-            such that you can quickly find the rules that contain a given feature.
-          - "hard rules" are those that contain substring/regex/bytes features or match statements.
-            these continue to be ordered topologically.
 
-        a rule evaluator can use the "easy rule" index to restrict the
-        candidate rules that might match a given set of features.
+        #
+        # Today, these scores are manually assigned by intuition/experience/guesswork.
+        # We could do a large-scale feature collection and use the results to assign scores.
+        #
 
-        at this time, a rule evaluator can't do anything special with
-        the "hard rules". it must still do a full top-down match of each
-        rule, in topological order.
+        if isinstance(
+            node,
+            capa.features.common.MatchedRule,
+        ):
+            # If present, other rule must match before this one, in same scope.
+            # Use score from that rule, which will have already been processed due to topological sorting.
+            # Otherwise, use a default score of 5.
+            assert isinstance(node.value, str)
+            return scores_by_rule.get(node.value, 5)
```

Comment:
What situation leads to a default value of `5` if the matched rule will have already been processed (and should have been scored)?

Reply:
Today, I think it's not possible. 

But this relies upon all scopes being processed in small->large order and the mapping to be correctly supplied. This wasn't always the case (in the development of the PR). 

So, I think we can remove this case in the code, though I feel this is a little brittle. I'll remove the branch and leave some further comments to explain where to look if there are future bugs.

Reply:
resolved in [457cfe0](https://github.com/mandiant/capa/pull/2080/commits/457cfe018053f224af9ad6681e767aee1268fa6c)

Reply:
I was wrong here and there's a CI failure due to this

Reply:
good this was caught

Reply:
An example of when this can happen today is here:

https://github.com/mandiant/capa-rules/blob/9e0ffdf7c51af31bb668c8ffbbe7c8f6fd9199cb/lib/allocate-or-change-rw-memory.yml#L24

`allocate or change rw memory` is a `static: basic block`/`dynamic: call` rule that has a `instruction` subscope block. This subscope block gets translated into a derived rule with `static: instruction` scope. When capa tries to score the `match: <subscope rule>` in dynamic mode, it can't find it, because `static: instruction` scoped rules won't ever run/be matched. So, we'd need a default score for this case (that will never be matched, so maybe it should be `score = 1`).

The most correct thing to do is to prune this impossible match in dynamic mode, but this requires some partial evaluation infrastructure that we don't have yet. That should be done in another issue.

Reply:
This did also uncover: https://github.com/mandiant/capa/issues/2124